### PR TITLE
Add more unit tests for ClientsManager.

### PR DESCRIPTION
**Pull request by upshaw-alex**
This PR adds additional unit tests for the `ClientsManager` class with the intention of reaching unit test coverage of as much of `ClientsManager`'s expected behaviors as possible.

Additionally, this commit also includes some minor fixes to `ClientsManager`'s implementation to address bugs in it that came up while writing these unit tests and some minor revisions to the documentation of `ClientManager`'s intended behavior in its header file to address certain ambiguities and inaccuracies in it that developing these unit tests highlighted.

### CODE REVIEW

**Comment by upshaw-alex** [comment]
I have force-pushed a rebase of my branch for this PR off of `master` in the interest of keeping this PR up-to-date.

**Comment by upshaw-alex** [comment]
I have force-pushed a minor amendment to my commit for this PR intended to address issues related to the clang-tidy check and compiler warnings/errors specific to certain build types.

**Review by evdzhurov — APPROVED** [review]
As discussed with @upshaw-alex there are a couple of unit tests in this PR that validate scenarios where some invariant of ClientsManager (CM) is broken.
- Setting the client batch size to 0 allows 1 reply to reside in reserved pages.
- Calling addPendingRequest multiple times can store more requests than the maximum number of requests per client.

These tests also helped identify possible misuse of the CM's interface. My recommendation is to omit those tests that I will mention with comments in the code (or convert them to a form that shows that the CM interface cannot be misused - these will fail however until improvements are made) and create a separate task with the following CM improvements (I can take this as an assignment when time permits).
- ClientsManager should not accept client batch size of 0 as a valid configuration (if not done already somewhere else).
- allocateNewReplyMsgAndWriteToStorage should assert properly if there's a space for insertion after deleteOldestReply. Currently the code only verifies that the current size is not greater than the max (but being equal to the max is also invalid for insertion).
- addPendingRequest should fail to add more requests and indicate an error if the size would exceed maxNumOfReqsPerClient_, this would help prevent programmer errors (calling addPendingReques twice or forgetting to call canBecomePending beforehand)
- fix removeRequestsOutOfBatchBounds  - TBD with @yontyon

**Review by teoparvanov — APPROVED** [review]
Thanks @upshaw-alex for creating the framework and the unit test, and to @evdzhurov for the detailed review. 

Alex, let's merge this once you remove the tests pointed out by Evgeniy. We will track fixing these corner cases in a separate tech debt work item.

**Comment by upshaw-alex** [comment]
I have force-pushed an amendment to my commit for this PR to adopt suggestions from @evdzhurov's review.

File: bftengine/tests/clientsManager/ClientsManager_test.cpp [hunk 1]
```diff
@@ -32,13 +249,1153 @@ TEST(ClientsManager, reservedPagesPerClient) {
...
+
+TEST(ClientsManager, numberOfRequiredReservedPagesSucceeds) {
+  resetMockReservedPages();
+  unique_ptr<ClientsManager> cm(new ClientsManager({1, 2, 3}, {4}, {5, 6}, metrics));
+
+  // Note we only check that numberOfRequiredReservedPages does not throw an exception or otherwise crash, and we do not
+  // actually check the values retruned by numberOfRequiredReservedPages, as exactly what value this function returns is
+  // considered an implementation detail of ClientsManager subject to change and not something guaranteed by
+  // ClientsManager's interface.
+  EXPECT_NO_THROW(cm->numberOfRequiredReservedPages()) << "ClientsManager::numberOfRequiredReservedPages failed.";
```
[the remark this answers is not in the document]
Comment (upshaw-alex) [reply bftengine/tests/clientsManager/ClientsManager_test.cpp]: I have adopted this suggestion to assert the number of required reserved pages returned is greater than `0` in a recent amendment to my commit for this PR.

File: bftengine/tests/clientsManager/ClientsManager_test.cpp [hunk 2]
```diff
@@ -32,13 +249,1153 @@ TEST(ClientsManager, reservedPagesPerClient) {
...
+                                         "ClientReplyMsg with a non-empty payload when called with an empty payload.";
+
+  string reply_1 = "reply 1 to client 2";
+  string reply_2 = "reply 2 to client 2";
+
+  ReplicaConfig::instance().setclientBatchingEnabled(true);
+  ReplicaConfig::instance().setclientBatchingMaxMsgsNbr(0);
+  cm.reset(new ClientsManager({}, {1, 2, 3}, {4}, metrics));
+
+  // Note that the influence of clientBatchingMaxMsgsNbr on ClientsManager::allocateNewReplyMsgAndWriteToStorage is that
```
[the remark this answers is not in the document]
Comment (upshaw-alex) [reply bftengine/tests/clientsManager/ClientsManager_test.cpp]: I have removed the part of this test case using a maximum client batch size of 0 in a recent amendment to my commit for this PR.

File: bftengine/tests/clientsManager/ClientsManager_test.cpp [hunk 3]
```diff
@@ -32,13 +249,1153 @@ TEST(ClientsManager, reservedPagesPerClient) {
...
+         "enabled.";
+
+  cm->removePendingForExecutionRequest(3, 3);
+  EXPECT_TRUE(cm->canBecomePending(3, 3))
+      << "ClientsManager::canBecomePending returned false for a recently removed request record whose removal should "
+         "have allowed canBecomePending to return true for it.";
+
+  cm->addPendingRequest(3, 3, "correlation ID");
+  cm->addPendingRequest(3, 4, "correlation ID");
+  EXPECT_TRUE(cm->canBecomePending(3, 5))
```
[the remark this answers is not in the document]
Comment (upshaw-alex) [reply bftengine/tests/clientsManager/ClientsManager_test.cpp]: I have removed this particular check and its associated setup in a recent amendment to my commit for this PR.

File: bftengine/tests/clientsManager/ClientsManager_test.cpp [hunk 4]
```diff
@@ -32,13 +249,1153 @@ TEST(ClientsManager, reservedPagesPerClient) {
...
+  ReplicaConfig::instance().setclientBatchingEnabled(false);
+  cm.reset(new ClientsManager({}, {2, 3, 5, 7}, {1}, metrics));
+  cm->addPendingRequest(3, 2, "correlation ID");
+  EXPECT_FALSE(cm->canBecomePending(3, 3))
+      << "ClientsManager::canBecomePending returned true for a client ID for which the ClientsManager should have "
+         "already had an existing request record with client batching disabled.";
+
+  cm->addPendingRequest(3, 3, "correlation ID");
+  EXPECT_TRUE(cm->canBecomePending(3, 4))
+      << "ClientsManager::canBecomePending returned false for a client ID/request sequence number combination for "
```
[the remark this answers is not in the document]
Comment (upshaw-alex) [reply bftengine/tests/clientsManager/ClientsManager_test.cpp]: I have removed this particular check and its associated setup in a recent amendment to my commit for this PR.

File: bftengine/tests/clientsManager/ClientsManager_test.cpp [hunk 5]
```diff
@@ -32,13 +249,1153 @@ TEST(ClientsManager, reservedPagesPerClient) {
...
+         "which there should have been no existing record and for a client for which it should have had more than 1 "
+         "existing request record with client batching disabled (note canBecomePending's check for a full batch is "
+         "only expected to recognize a number of existing records exactly equal to the effective maximum client batch "
+         "size, not numbers of records exceeding it (and that the effective maximum client batch size is 1 when client "
+         "batching is disabled)).";
+
+  ReplicaConfig::instance().setclientBatchingEnabled(true);
+  ReplicaConfig::instance().setclientBatchingMaxMsgsNbr(0);
+  cm.reset(new ClientsManager({}, {2, 3, 5, 7}, {1}, metrics));
+  EXPECT_FALSE(cm->canBecomePending(3, 1))
```
[the remark this answers is not in the document]
Comment (upshaw-alex) [reply bftengine/tests/clientsManager/ClientsManager_test.cpp]: I have entirely removed the portion of this test case that uses a maximum client batch size of `0` from this test case based on your recommendation that a client batch size of `0` doesn't make sense and that we should move to reject it in the future.

File: bftengine/tests/clientsManager/ClientsManager_test.cpp [hunk 6]
```diff
@@ -32,13 +249,1153 @@ TEST(ClientsManager, reservedPagesPerClient) {
...
+      << "ClientsManager::canBecomePending returned true for a request with a client ID for which the ClientsManager "
+         "should have had no existing request records with client batching enabled and a maximum client batch size of "
+         "0.";
+  cm->addPendingRequest(3, 1, "correlation ID");
```
[the remark this answers is not in the document]
Comment (upshaw-alex) [reply bftengine/tests/clientsManager/ClientsManager_test.cpp]: I have entirely removed the portion of this test case that uses a maximum client batch size of `0` from this test case based on your recommendation that a client batch size of `0` doesn't make sense and that we should move to reject it in the future.

File: bftengine/tests/clientsManager/ClientsManager_test.cpp [hunk 7]
```diff
@@ -32,13 +249,1153 @@ TEST(ClientsManager, reservedPagesPerClient) {
...
+
+  cm->removePendingForExecutionRequest(8, maxNumOfRequestsInBatch - 1);
+  cm->removeRequestsOutOfBatchBounds(8, maxNumOfRequestsInBatch - 1);
+  EXPECT_TRUE(cm->isClientRequestInProcess(8, maxNumOfRequestsInBatch))
+      << "ClientsManager::removeRequestsOutOfBatchBounds removed the request record with the greatest sequence number "
+         "when it should have had fewer than maxNumOfRequestsInBatch request records.";
+  cm->addPendingRequest(8, maxNumOfRequestsInBatch + 1, "correlation ID");
+  cm->addPendingRequest(8, maxNumOfRequestsInBatch + 2, "correlation ID");
+  cm->removeRequestsOutOfBatchBounds(8, maxNumOfRequestsInBatch - 1);
+  EXPECT_TRUE(cm->isClientRequestInProcess(8, maxNumOfRequestsInBatch + 2))
```
[the remark this answers is not in the document]
Comment (upshaw-alex) [reply bftengine/tests/clientsManager/ClientsManager_test.cpp]: I have removed this particular check and its associated setup in a recent amendment to my commit for this PR since I went ahead and did so in other cases where the tests were relying on `addPendingRequest` to not enforce limits we think it probably should be enforcing going forward.

I also thought it was odd `removeRequestsOutOfBatchBounds` uses the global system constant `maxNumOfRequestsInBatch` instead of the effective maximum client batch size, and I would be curious to hear if this is found to have been intentional or not.

**Comment by upshaw-alex** [comment]
I have force-pushed a rebase of my branch for this PR off of `master` in the interest of keeping this PR up-to-date.

File: bftengine/tests/clientsManager/ClientsManager_test.cpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (evdzhurov) [reply bftengine/tests/clientsManager/ClientsManager_test.cpp]: Hey, I spoke with @yontyon about removeRequestsOutOfBatchBounds. It appears that maxNumOfReqsPerClient_ should've been used, but it's best if he confirms it 100%. The implication here is that because the global constant is probably larger than maxNumOfReqsPerClient_, removeRequestsOutOfBatchBounds might not have been able to have any effect. Meaning that fixing this method now is subject to testing.

**Comment by upshaw-alex** [comment]
I have force-pushed an amendment to my commit for this PR to rebase it off of `master`. I was originally inspired to make this rebase in the general interest of keeping this PR up-to-date and in the specific interest of getting some changes to disable tests that are currently failing on `master` since I have been getting some test failures on this PR that look unrelated to its changes.

In the process of rebasing, I discovered merge conflicts with https://github.com/vmware/concord-bft/pull/2127, which makes changes to `ClientsManager`'s interface and behavior. I have attempted to include a resolution to these merge conflicts in my rebase commit, but I have some uncertainties about them since I have some questions about exactly what changes https://github.com/vmware/concord-bft/pull/2127 intended to make to `ClientsManager`'s behavior. I have raised these questions in a comment on that MR (https://github.com/vmware/concord-bft/pull/2127#issuecomment-996350769). My current resolution to the merge conflicts changes assumes that https://github.com/vmware/concord-bft/pull/2127 's overloading of `ClientsManager`'s constructor was intentional and that the fact that the original version of the constructor without the new `std::shared_ptr<PersistentStorage> ps` parameter was left `public` is intentional, but that it is not intentional that `ClientsManager`s so constructed segfault when `ClientsManager::allocateNewReplyMsgAndWriteToStorage` and/or `ClientsManager::allocateReplyFromSavedOne` are called. I have chosen to start with these particular assumptions for now because they yield the simplest case for resolving the merge conflicts with the tests I am trying to commit here.
